feat: add `autogram.Engine` (batched) (#387)
TorchJD/torchjd: docs/source/examples/iwrm.rst

This example shows how to use TorchJD to minimize the vector of per-instance losses. This learning
paradigm, called IWRM, is multi-objective, as opposed to the usual empirical risk minimization
(ERM), which seeks to minimize the average loss. While a step of ERM may increase the loss of some
samples of the batch, a step of IWRM guarantees that no loss from the batch is increased (given a

> sufficiently small learning rate).

Reviewer: This is specific to `UPGrad`, other aggregators may do whatever they want.

.. hint::
    A proper definition of IWRM and its empirical results on some deep learning tasks are
    available in `Jacobian Descent For Multi-Objective Optimization
    <https://arxiv.org/pdf/2406.16232>`_.

TorchJD offers two methods to perform IWRM. The :doc:`autojac <../docs/autojac/index>` engine
backpropagates the Jacobian of each sample's loss. It uses an
:doc:`Aggregator <../docs/aggregation/index>` to combine the rows of this Jacobian to fill the

> ``.grad`` fields of the model's parameters. Because it has to store the full Jacobian in memory,
> this approach consumes a lot of memory.

Author: ```suggestion
``.grad`` fields of the model's parameters. Because it has to store the full Jacobian, this approach
uses a lot of memory.
```

                optimizer.zero_grad()
                loss.backward()

                optimizer.step()

        In this baseline example, the update may negatively affect the loss of some elements of the
        batch.

    .. tab-item:: autojac

        .. code-block:: python
            :emphasize-lines: 5-6, 12, 16, 21, 23

            import torch
            from torch.nn import Linear, MSELoss, ReLU, Sequential
            from torch.optim import SGD

            from torchjd.aggregation import UPGrad
            from torchjd.autojac import backward

            X = torch.randn(8, 16, 10)
            Y = torch.randn(8, 16)

            model = Sequential(Linear(10, 5), ReLU(), Linear(5, 1))
            loss_fn = MSELoss(reduction="none")

            params = model.parameters()
            optimizer = SGD(params, lr=0.1)
            aggregator = UPGrad()

            for x, y in zip(X, Y):
                y_hat = model(x).squeeze(dim=1)  # shape: [16]
                losses = loss_fn(y_hat, y)  # shape: [16]
                optimizer.zero_grad()
                backward(losses, aggregator)

                optimizer.step()

>         Here, we compute the Jacobian of per-sample losses with respect to the model parameters and

Author: ```suggestion
        Here, we compute the Jacobian of the per-sample losses with respect to the model parameters and
```

>         use it to update the model such that no loss from the batch is locally negatively affected.

Author: ```suggestion
        use it to update the model such that no loss from the batch is (locally) negatively affected.
```
